### PR TITLE
Parsing NewTarget meta property, closes #37

### LINKED ISSUE #37: new.target support
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new.target

### PR DESCRIPTION
**Pull request by cmtt**
Parses `new.target` (#37).

Notes:
- ~~~~~`MetaPropertyExpression` has been used instead of MetaProperty~~~~~
- No checks whether the token appears only inside Function bodies.
- only `NewTarget` is a valid MetaProperty, thus, `property` is compared with `"target"` in `parser::meta_property_expression`

### CODE REVIEW

File: ratel-codegen/src/expression.rs [hunk 1]
```diff
@@ -18,6 +18,7 @@ impl<'ast, G: Generator> ToCode<G> for Expression<'ast> {
             Array(ref array)             => gen.write(array),
             Member(ref member)           => gen.write(member),
             ComputedMember(ref computed) => gen.write(computed),
+            MetaMember(ref property)     => gen.write(property),
```
Comment (maciejhirsz) [inline ratel-codegen/src/expression.rs]: Should be `MetaProperty`

File: ratel-visitor/src/expression.rs [hunk 2]
```diff
@@ -42,6 +42,10 @@ impl<'ast> Visitable<'ast> for ExpressionNode<'ast> {
                 computed.traverse(visitor, ctx);
                 visitor.on_computed_member_expression(computed, self, ctx);
             },
+            MetaMember(ref property) => {
```
Comment (maciejhirsz) [inline ratel-visitor/src/expression.rs]: Should be `MetaProperty`

File: ratel-visitor/src/lib.rs [hunk 3]
```diff
@@ -49,6 +49,8 @@ build! {
     fn on_array_expression(item: &ArrayExpression<'ast>, node: &ExpressionNode<'ast>);
     fn on_member_expression(item: &MemberExpression<'ast>, node: &ExpressionNode<'ast>);
     fn on_computed_member_expression(item: &ComputedMemberExpression<'ast>, node: &ExpressionNode<'ast>);
+    fn on_meta_member_expression(item: &MetaPropertyExpression<'ast>, node: &ExpressionNode<'ast>);
```
Comment (maciejhirsz) [inline ratel-visitor/src/lib.rs]: Should be `on_meta_proprety_expression`

File: ratel/src/ast/expression.rs [hunk 4]
```diff
@@ -160,6 +167,7 @@ impl_from! {
     ArrayExpression<'ast> => Array,
     MemberExpression<'ast> => Member,
     ComputedMemberExpression<'ast> => ComputedMember,
+    MetaPropertyExpression<'ast> => MetaMember,
```
Comment (maciejhirsz) [inline ratel/src/ast/expression.rs]: `MetaProperty`

**Comment by maciejhirsz** [comment]
Also, a comment regarding this:

> No checks whether the token appears only inside Function bodies.

if the special case for handling the `"new"` token is put into it's own handler that can be fixed in the lookup tables, it's possible to do context switches in parser using generics at 0 cost. I've also experimented with it in Lunarity (it's not using lookup tables here, but you get the idea):
https://github.com/paritytech/lunarity/blob/5d4cb5ce976f560371eec7156574000b22314fe8/core/src/parser/statement.rs#L7-L39

**Comment by cmtt** [comment]
Context switches as implemented in lunarity for restricting according to grammar seem to be a great addition. This would also help to add more meaningful parser errors in the long term.

- In the long run, I share the thought that adding enums for upcoming other meta properties would be beneficial. As far as the ES6 [spec](http://www.ecma-international.org/ecma-262/6.0/#sec-meta-properties) is concerned, I can only see that NewTarget is the only currently allowed meta property.
- The other changes have been applied.
- The `NEW` handler has been added. According to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence), the precedence of `new` without arguments differs from `new` with arguments.

File: ratel-visitor/src/expression.rs [hunk 5]
```diff
@@ -52,6 +52,10 @@ impl<'ast> Visitable<'ast> for ExpressionNode<'ast> {
                 visitor.push_parent(ParentNode::from(self), ctx);
                 computed.traverse(visitor, ctx);
             },
+            MetaProperty(ref property) => {
+                property.traverse(visitor, ctx);
+                visitor.on_meta_property(property, self, ctx);
```
Comment (maciejhirsz) [inline ratel-visitor/src/expression.rs]: After the parent tracking PR, the order here should be:

1) `on_meta_property`
2) `push_parent` (new)
3) `traverse`

File: ratel/src/astgen/expression.rs [hunk 6]
```diff
@@ -74,6 +74,18 @@ impl<'ast> SerializeInLoc for ComputedMemberExpression<'ast> {
     }
 }
 
+impl<'ast> SerializeInLoc for MetaPropertyExpression<'ast> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::SerializeStruct, S::Error>
+    where
+        S: Serializer,
+    {
+        self.in_loc(serializer, "MetaProperty", 2, |state| {
+                state.serialize_field("meta", &self.meta)?;
```
Comment (maciejhirsz) [inline ratel/src/astgen/expression.rs]: weird indentation here :upside_down_face:

File: ratel/src/parser/expression.rs [hunk 7]
```diff
@@ -141,6 +141,25 @@ create_handlers! {
         par.alloc_at_loc(start, end, expression)
     };
 
+    pub const NEW = |par| {
+        let (start, op_end) = par.lexer.loc();
+        let op = OperatorKind::from_token(par.lexer.token).expect("Must be a prefix operator");
```
Comment (maciejhirsz) [inline ratel/src/parser/expression.rs]: The check and the variable are unnecessary, you are guaranteed that the token is `NewOperator` now :)

File: ratel/src/parser/expression.rs [hunk 8]
```diff
@@ -141,6 +141,25 @@ create_handlers! {
...
+
+        par.lexer.consume();
+
+        if par.lexer.token == Accessor {
+            let meta = par.alloc_at_loc(start, op_end, op.as_str());
```
Comment (maciejhirsz) [inline ratel/src/parser/expression.rs]: `op.as_str()` to `"new"`

File: ratel/src/parser/expression.rs [hunk 9]
```diff
@@ -141,6 +141,25 @@ create_handlers! {
...
+            let expression = par.meta_property_expression(meta);
+            let end = par.lexer.end();
+            par.lexer.consume();
+            par.alloc_at_loc(start, end, expression)
+        } else {
+            let expression = par.prefix_expression(op);
```
Comment (maciejhirsz) [inline ratel/src/parser/expression.rs]: `op` to `OperatorKind::New`

**Pull request merged**